### ISSUE TITLE
Parsing postgres connection strings query parameters

### ISSUE THREAD

**Issue by elhigu** [tgriesser/knex#2354]
Knex version: 0.14.2
Database + version: pg
OS: all

# Feature discussion / request

Currently postgres query string parser doesn't include all query parameters automatically to the connection object. I haven't checked the current code, but according to #1738
`postgresql://user:pass@host/dbname?application_name=app-name` parameter is not passed to the driver.

I suppose we could use some standard url parsing library (if not used already) and add all the parsed query parameters to the connection object.

**Comment by vitaly-t** [comment]
If this issue is still valid, there's generic [connection-string] now that can be used with any database engine.

[connection-string]:https://github.com/vitaly-t/connection-string

**Comment by elhigu** [comment]
@vitaly-t that did seem promising. I think we could use it somehow like: 

```
{
  client: 'pg', 
  genericConnection: 'vitaly://syntax@connection ',
 pool: {...}
}
```

and connection object would be populated with values from `genericConnection` string...

I'm not sure, but each dialect might need to modify result got from that parser though... If there is no need to modify generic parser result we could just use it directly like: 

```
{
  client: 'pg', 
  connection: parse('vitaly://syntax@connection'),
 pool: {...}
}
```

and make no changes to knex.

**Comment by vitaly-t** [comment]
@elhigu Yes, you would want to use some form of a simple adapter, to account for specifics of the underlying driver, as I outlined on the [Adapters](https://github.com/vitaly-t/connection-string/wiki/Adapters) page.

**Comment by elhigu** [comment]
Would it make sense to create repo to provide also pre-written adapters for all the databases in the world or should every project copy paste them in from wiki? (I'm not sure myself either which would be better way...)

**Comment by vitaly-t** [comment]
And as for the syntax example, I would think that the most logical approach is to use the protocol to tell which driver should be used:

```
postgresql://...
//=> use node-postgres driver

mysql://...
//=> use mysql driver

etc.
```

So there is no need to specify the client in a separate `client` property.

**Comment by vitaly-t** [comment]
> Would it make sense to create repo to provide also pre-written adapters for all the databases in the world or should every project copy paste them in from wiki?

I cannot generalize here. For each driver this needs to be looked into separately, to see what specifics, if any it supports, and work-around those. The online examples give you the general idea. But some drivers may support other non-standard things, that's their problem.

**Comment by elhigu** [comment]
Or maybe just write a separate package which takes in connection string and driver name and converts them to connection object for the given driver library.

I know it would be logical, but for example mysql has 2 different drivers for the same protocol right now... It is possible too to replace protocol with driver name, except some drivers then accept multiple protocols in 🤔 So maybe safest to have separate client attribute in addition to connection attributes.

**Comment by vitaly-t** [comment]
> Or maybe just write a separate package which takes in connection string and driver name and converts them to connection object for the given driver library.

In case of any integration library, like knex or sequelize, this is exactly what makes sense! :smile:

**Comment by elhigu** [comment]
Lets do it like that, so there won't be need to made any other changes to knex except to document how to use that parser and promote the library 👍 That will also make dealing with connection strings in knex-db-manager easier :) For now it has been only accepting connection objects.

One drawback in that approach is that there would not be guarantees that username, password, database, port and host will be found from same attributes in knex's connection object, when we are inspecting connection details. Though I'm not sure if there has been any guarantees for those earlier either.

**Comment by vitaly-t** [comment]
> One drawback in that approach is that there would not be guarantees that username, password, database, port and host will be found from same attributes in knex's connection object, when we are inspecting connection details.

That's what generic [connection-string] unifies. One step toward less ambiguity :smile:

[connection-string]:https://github.com/vitaly-t/connection-string

**Comment by elhigu** [comment]
It would require changing knex internals in a way that it would always have that generic object somewhere available and implementing converters from propietary format to generic to allow passing driver options in their native format (and if passed in native string format, then we would still not have generic object unless implementing parsers for native connection strings).

I think that supporting drivers pure configs perfectly is more important than having the generic config support. Both is better of course :)

**Comment by vitaly-t** [comment]
Yes, I think you would have generic connection string protocol in the front, from which you then pull values into custom connection objects, according to the underlying driver requirements.

I was just distracted, working on another update, which finally has rectified how Unix sockets should be handled. From [here](https://github.com/vitaly-t/connection-string/wiki/FAQ):

> **Q:** How can I specify a Unix-socket host?

> **A:** Any host name that either ends with `.sock` or contains `/`, after decoding, is recognized as a Unix socket,
with the host's `type` property set to `socket` after parsing.

> Note that every symbol `/`, in a socket name inside URL, must be encoded as `%2F`.

**Comment by AndreiIgna** [comment]
Just encountered a similar problem with MySQL.

Until now I was using `mysql` lib directly with a connection string like this one `mysql://user:pass@localhost/db?socketPath=/cloudsql/project:region:dbname`
This was fine until I switched to using Knex, with same connection string. The config parser Knex doesn't use the query strings, so the connection to database fails. 

I've checked the implementation in both libraries, and the parsing result looks different:
With Knex:
```js
{ client: 'mysql',
  connection:
   { database: 'db',
     host: 'localhost',
     user: 'user',
     password: 'pass'  } }
// socketPath missing!
```
With mysql:
```js
{ database: 'db',
  host: 'localhost',
  user: 'user',
  password: 'pass',
  socketPath: '/cloudsql/project:region:dbname' }
```

Looks like the query string params can be appended easily to config object and it will be helpful for more connection drivers. Is it ok to make a PR for this? Similar with the solution here https://github.com/mysqljs/mysql/blob/master/lib/ConnectionConfig.js#L186

Cheers

**Comment by mercmobily** [comment]
Hello! Thanks for reporting this issue.

This issue has been resolved. Query parameters in connection strings are now properly parsed and passed through to the database drivers for all supported databases.

The fix was implemented in commit #3547 by AndreiIgna on 21 November 2019, shortly after you opened this issue. The implementation parses query string parameters from connection URLs and automatically adds them to the connection object, which is then passed to the underlying database driver.

I've tested the current version (3.1.0) and can confirm that all the use cases mentioned in this issue now work correctly:

**PostgreSQL with application_name** (your original issue):
```javascript
parseConnection('postgresql://user:pass@host/dbname?application_name=app-name')
// Returns connection object with application_name: 'app-name' ✓
```

**MySQL with socketPath** (AndreiIgna's use case):
```javascript
parseConnection('mysql://user:pass@localhost/db?socketPath=/cloudsql/project:region:dbname')
// Returns connection object with socketPath: '/cloudsql/project:region:dbname' ✓
```

**Multiple query parameters**:
```javascript
parseConnection('mysql://user:pass@localhost/db?socketPath=/tmp/mysql.sock&charset=utf8mb4')
// Returns connection object with both socketPath and charset parameters ✓
```

The implementation at `lib/knex-builder/internal/parse-connection.js:70-85` now iterates through all query parameters using the URL `searchParams` API and adds them to the connection object. For MySQL, MariaDB, and MSSQL, there's additional support for JSON-encoded parameter values (added in #4629), which allows passing complex nested configuration options through the connection string.

The solution works for all database dialects, though PostgreSQL connections are handled slightly differently as they use the `pg-connection-string` library, which has its own query parameter parsing implementation. This means PostgreSQL query parameters have been working since that library added support for them.

There is a related broader issue at #1528 which discusses connection string query parameter parsing across all databases and has received more detailed analysis from the maintainers. That issue also mentions the potential future adoption of the generic `connection-string` library (discussed in #2720), which would further standardise connection string handling across all database types.

This issue can be closed as resolved in current versions.


## Related Issues

- #1528 - knex doesn't parse connection URL query params into connection object (broader discussion, opened 2016)
- #1738 - application name for postgresql (closed, specific to PostgreSQL application_name parameter)
- #1568 - using connection DSN doesn't support sockets (closed, related to socketPath)
- #2720 - Use connection-string library for generic connection string (open, discusses future improvements)
